tsl/src/continuous_aggs/refresh.c

```diff
@@ -907,3 +924,240 @@ continuous_agg_refresh_internal(const ContinuousAgg *cagg,
+				hypertable_id = $1 \
+		) \
+		SELECT \
+			refresh_start AS start, \
+			LEAST($6::numeric, refresh_start::numeric + $4::numeric)::bigint AS end \
+		FROM \
+			pg_catalog.generate_series($5, $6, $4) AS refresh_start \
+		WHERE \
+			EXISTS ( \
+			    SELECT FROM chunk_ranges \
+				WHERE \
+					pg_catalog.int8range(refresh_start, LEAST($6::numeric, refresh_start::numeric + $4::numeric)::bigint) \
+					OPERATOR(pg_catalog.&&) \
+					pg_catalog.int8range(chunk_ranges.start, chunk_ranges.end) \
+			) \
+			AND EXISTS ( \
+				SELECT FROM \
+					invalidation_logs \
+				WHERE \
+					pg_catalog.int8range(refresh_start, LEAST($6::numeric, refresh_start::numeric + $4::numeric)::bigint) \
+					OPERATOR(pg_catalog.&&) \
+					pg_catalog.int8range(lowest_modified_value, greatest_modified_value) \
+					AND lowest_modified_value IS NOT NULL \
+					AND (greatest_modified_value IS NOT NULL AND greatest_modified_value != -210866803200000001) \
+			) \
+		ORDER BY \
+			refresh_start DESC;";
+
+	List *refresh_window_list = NIL;
+	int res;
+	Oid types[] = { INT4OID, INT4OID, INT4OID, INT8OID, INT8OID, INT8OID };
+	Datum values[] = { Int32GetDatum(ht->fd.id),
+					   Int32GetDatum(time_dim->fd.id),
+					   Int32GetDatum(cagg->data.mat_hypertable_id),
+					   Int64GetDatum(batch_size),
+					   Int64GetDatum(refresh_window.start),
+					   Int64GetDatum(refresh_window.end) };
+	char nulls[] = { false, false, false, false, false, false };
+	MemoryContext oldcontext = CurrentMemoryContext;
+
+	/*
+	 * Query for the oldest chunk in the hypertable.
+	 */
+	if (SPI_connect() != SPI_OK_CONNECT)
+		elog(ERROR, "could not connect to SPI");
+
+	res = SPI_execute_with_args(query_str,
+								6,
+								types,
+								values,
+								nulls,
+								false /* read_only */,
+								0 /* count */);
+
+	if (res < 0)
+		elog(ERROR, "%s: could not get the last bucket of the materialized data", __func__);
+
+	for (uint64 i = 0; i < SPI_processed; i++)
+	{
+		bool range_start_isnull, range_end_isnull;
+		Datum range_start =
+			SPI_getbinval(SPI_tuptable->vals[i], SPI_tuptable->tupdesc, 1, &range_start_isnull);
+		Datum range_end =
+			SPI_getbinval(SPI_tuptable->vals[i], SPI_tuptable->tupdesc, 2, &range_end_isnull);
+
+		/* We need to allocate the list in the old memory context because here we're in the SPI
+		 * context */
+		MemoryContext saved_context = MemoryContextSwitchTo(oldcontext);
+		InternalTimeRange *range = palloc0(sizeof(InternalTimeRange));
+		range->start = DatumGetInt64(range_start);
+		range->start_isnull = range_start_isnull;
+		range->end = DatumGetInt64(range_end);
+		range->end_isnull = range_end_isnull;
+		range->type = original_refresh_window->type;
+
+		/* When dropping chunks we need to align the start of the first range to cover dropped
+		 * chunks if they exist */
+		if (i == (SPI_processed - 1) && original_refresh_window->start_isnull)
+		{
+			range->start = original_refresh_window->start;
+			range->start_isnull = true;
```

Comment:
Isn't the invariant that `range->start == DT_NOBEGIN` if and only if `range->start_isnull == true`? So, shouldn't you set `range->start` to `DT_NOBEGIN`? I can imagine a situation where the original refresh window is not `DT_NOBEGIN` but you call some code that assume the invariant above and end up with a bad computation. For example, `ts_compute_inscribed_bucketed_refresh_window_variable` relies on this.

Reply:
Fixed